### ISSUE TITLE
Most database platforms do not lock after poll

### ISSUE THREAD

**Issue by roelvanduijnhoven** [juriansluiman/SlmQueueDoctrine#128]
Whenever I spin up multiple workers I get the following error every now and then:
> Race-condition detected while updating item in queue

This effect is easily reproducible:
1. Spin up multiple workers
2. Remove the `IdleNapStrategy` (optional, but in this case it will directly crash)
3. Push some jobs on the queue

The problem is that for almost all database platforms no locking is used at all. That is due to the way we call DBAL:

> $queryBuilder->from($platform->appendLockHint($this->options->getTableName(), LockMode::PESSIMISTIC_READ));

This will hint to use locking. However most platforms will not do anything with it. In my case I am using MariaDB which uses the MySql platform. And that does nothing with this hint:

```sql
START TRANSACTION;
SELECT * FROM queue_default WHERE (status = ?) AND (queue = ?) AND (scheduled <= ?) ORDER BY priority ASC, scheduled ASC LIMIT 1;
COMMIT;
```

We can fix this by using Doctrine's ORM functionallity to properly lock. I do not see an easy way to implement this without ORM; as DBAL does not contain functionality for it. 

What are your thoughts on this @basz?

**Comment by basz** [comment]
If it can be easily fixed by using ORM I see no real reason not to use it. I know the tendency in this project has been to prefer DBAL over full ORM but, if we can solve it generically with it, why not?

How many worker do you use? I never have seen this. Also I'm no lock expert so I assume you know what we're talking about exactly ;-)

**Comment by roelvanduijnhoven** [comment]
@basz What database engine are you using? I experience it on MariaDb with only 2 workers. It does happen more easily when the jobs tend to be very small (or aborted early due to rate limiting in my case); and thus the poll query is executed more often.

I'll try to properly implement locking then. Not an expert myself as well; but let's see what I end up with ;).

**Comment by basz** [comment]
MySQL with 1 to 3 workers over a dozen projects.

**Comment by basz** [comment]
closed via #129
